lib/queue/job.js

```diff
@@ -788,29 +805,66 @@ Job.prototype.save = function( fn ) {
   // update
   if( this.id ) return this.update(fn);
 
+  // unique id?
+  if( this._unique ) return this.saveUnique(fn);
+
   // incr id
   client.incr(client.getKey('ids'), function( err, id ) {
     if( err ) return fn(err);
-    // add the job for event mapping
-    self.id = id;
-    self.zid = client.createFIFO(id);
-    self.subscribe(function() {
-      self._state     = self._state || (this._delay ? 'delayed' : 'inactive');
-      if( max ) { self.set('max_attempts', max); }
-      client.sadd(client.getKey('job:types'), self.type, noop);
-      self.set('type', self.type);
-      var now         = Date.now();
-      self.created_at = now;
-      self.set('created_at', self.created_at);
-      self.promote_at = now + (self._delay || 0);
-      self.set('promote_at', self.promote_at);
-      self.update(fn);
-    }.bind(this));
+    self.saveWithId(id, fn);
   }.bind(this));
   return this;
 };
 
 /**
+ * Save the job, optionally invoking the callback `fn(err)`.
+ *
+ * @param {number} id
+ * @param {Function} fn
+ * @return {Job} for chaining
+ * @api public
+ */
+
+Job.prototype.saveUnique = function( id, fn ) {
+  throw new Error('TODO: Check/set a hash, mapping unique identifiers => integer id');
```

Comment:
:question: It seems like I'll need to map the unique values (string) to an ID, grabbed by the same incrementing approach #save() uses.

I'm thinking it can be an HMAP at `ids:unique`

Steps:
1. HGET `ids:unique` `this._unique`
  If present, return noop
2. Use `client.incr(client.getKey('ids'), ...` to get next ID
3. HSET `ids:unique` `this._unique` `id`
4. call `saveWithId(id)`   (this is a new refactor I'm introducing in this PR)

I'm worried about race conditions with what I've described above